src/relax/utils.cc

```diff
@@ -111,6 +113,75 @@ bool IsLeafOrTuple(const Expr& expr) {
          expr.as<OpNode>() || expr.as<TupleNode>();
 }
 
+bool IsImpureCall(const Call& call) {
+  if (auto op_ptr = call->op.as<OpNode>()) {
+    auto op = GetRef<Op>(op_ptr);
+    static auto purity_map = Op::GetAttrMap<Bool>("FPurity");
+    ICHECK(purity_map.count(op)) << "Cannot find the registered purity of this op: " << op->name;
+    return !(purity_map[op]->value);
+  }
+  // the StructInfo must be FuncStructInfo
+  auto func_struct_info = GetStructInfoAs<FuncStructInfoNode>(call->op);
+  return !func_struct_info->purity;
+}
+
+Call WrapCallPure(const Call& call) {
```

Comment:
Same here, once we have special purity deduction, per op, likely we don;t need to rely on wrapping unwrapping in most cases